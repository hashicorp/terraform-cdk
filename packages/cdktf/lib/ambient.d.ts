// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
declare module 'archiver';