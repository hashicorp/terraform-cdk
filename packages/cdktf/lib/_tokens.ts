// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, DefaultTokenResolver, StringConcat } from "./tokens";
import { Construct } from "constructs";

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

// eslint-disable-next-line jsdoc/require-jsdoc
export function resolve<T>(scope: Construct, obj: T, preparing = false): T {
  return Tokenization.resolve(obj, {
    scope,
    preparing,
    resolver: TOKEN_RESOLVER,
  });
}
