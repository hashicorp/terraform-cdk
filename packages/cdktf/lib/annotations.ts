// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
// Originally from aws-cdk v2 because with constructs v10 addMessage was removed
// in favor of Annotations in the AWS CDK
// https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/annotations.ts
import { IConstruct } from "constructs";

export enum AnnotationMetadataEntryType {
  INFO = "@cdktf/info",
  WARN = "@cdktf/warn",
  ERROR = "@cdktf/error",
}

export const DISABLE_STACK_TRACE_IN_METADATA =
  "@cdktf/disable_stack_trace_in_metadata";

/**
 * Includes API for attaching annotations such as warning messages to constructs.
 */
export class Annotations {
  /**
   * Returns the annotations API for a construct scope.
   * @param scope The scope
   */
  public static of(scope: IConstruct) {
    return new Annotations(scope);
  }

  private readonly stackTraces: boolean;

  private constructor(private readonly scope: IConstruct) {
    const disableTrace =
      scope.node.tryGetContext(DISABLE_STACK_TRACE_IN_METADATA) || false;

    this.stackTraces = !disableTrace;
  }

  /**
   * Adds a warning metadata entry to this construct.
   *
   * The CLI will display the warning when an app is synthesized.
   * In a future release the CLI might introduce a --strict flag which
   * will then fail the synthesis if it encounters a warning.
   *
   * @param message The warning message.
   */
  public addWarning(message: string) {
    this.addMessage(AnnotationMetadataEntryType.WARN, message);
  }

  /**
   * Adds an info metadata entry to this construct.
   *
   * The CLI will display the info message when apps are synthesized.
   *
   * @param message The info message.
   */
  public addInfo(message: string): void {
    this.addMessage(AnnotationMetadataEntryType.INFO, message);
  }

  /**
   * Adds an { "error": <message> } metadata entry to this construct.
   * The toolkit will fail synthesis when errors are reported.
   * @param message The error message.
   */
  public addError(message: string) {
    this.addMessage(AnnotationMetadataEntryType.ERROR, message);
  }

  /**
   * Adds a message metadata entry to the construct node, to be displayed by the CDK CLI.
   * @param level The message level
   * @param message The message itself
   */
  private addMessage(level: AnnotationMetadataEntryType, message: string) {
    this.scope.node.addMetadata(level, message, {
      stackTrace: this.stackTraces,
    });
  }
}
