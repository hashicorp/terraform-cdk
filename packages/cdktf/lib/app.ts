import { Construct, Node, ConstructMetadata } from 'constructs';
import fs = require('fs');
import { version } from '../package.json';

export interface AppOptions {
    /**
     * The directory to output Terraform resources.
     *
     * @default - CDKTF_OUTDIR if defined, otherwise "cdktf.out"
     */
    readonly outdir?: string;
    readonly stackTraces?: boolean;
}

/**
 * Represents a cdktf application.
 */
export class App extends Construct {
    /**
     * The output directory into which resources will be synthesized.
     */
    public readonly outdir: string;

    /**
     * Defines an app
     * @param options configuration options
     */
    constructor(options: AppOptions = {}) {
        super(undefined as any, '');
        this.outdir = process.env.CDKTF_OUTDIR ?? options.outdir ?? 'cdktf.out';

        const node = Node.of(this)
        if (options.stackTraces === false) {
            node.setContext(ConstructMetadata.DISABLE_STACK_TRACE_IN_METADATA, true);
        }

        node.setContext('cdktfVersion', version)
    }

    /**
     * Synthesizes all resources to the output directory
     */
    public synth(): void {
        if (!fs.existsSync(this.outdir)) {
            fs.mkdirSync(this.outdir);
        }

        Node.of(this).synthesize({
            outdir: this.outdir
        });
    }
}