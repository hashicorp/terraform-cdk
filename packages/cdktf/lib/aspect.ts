// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
//  Originally from aws-cdk v2 because with constructs v10 Aspects where moved to the AWS CDK
//  https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/aspect.ts
import { IConstruct } from "constructs";

const ASPECTS_SYMBOL = Symbol("cdktf-aspects");

/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * All aspects can visit an IConstruct
   */
  visit(node: IConstruct): void;
}

/**
 * Aspects can be applied to CDK tree scopes and can operate on the tree before
 * synthesis.
 */
export class Aspects {
  /**
   * Returns the `Aspects` object associated with a construct scope.
   * @param scope The scope for which these aspects will apply.
   */
  public static of(scope: IConstruct): Aspects {
    let aspects = (scope as any)[ASPECTS_SYMBOL];
    if (!aspects) {
      aspects = new Aspects();

      Object.defineProperty(scope, ASPECTS_SYMBOL, {
        value: aspects,
        configurable: false,
        enumerable: false,
      });
    }
    return aspects;
  }

  private readonly _aspects: IAspect[];

  private constructor() {
    this._aspects = [];
  }

  /**
   * Adds an aspect to apply this scope before synthesis.
   * @param aspect The aspect to add.
   */
  public add(aspect: IAspect) {
    this._aspects.push(aspect);
  }

  /**
   * The list of aspects which were directly applied on this scope.
   */
  public get all(): IAspect[] {
    return [...this._aspects];
  }
}
