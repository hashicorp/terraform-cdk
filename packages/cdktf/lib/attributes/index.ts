export * from "./terraform-any-attribute"
export * from "./terraform-any-list-attribute"
export * from "./terraform-any-map-attribute"
export * from "./terraform-any-set-attribute"
export * from "./terraform-attribute"
export * from "./terraform-boolean-attribute"
export * from "./terraform-boolean-list-attribute"
export * from "./terraform-boolean-map-attribute"
export * from "./terraform-boolean-set-attribute"
export * from "./terraform-dynamic-attribute"
export * from "./terraform-list-attribute"
export * from "./terraform-map-attribute"
export * from "./terraform-number-attribute"
export * from "./terraform-number-list-attribute"
export * from "./terraform-number-map-attribute"
export * from "./terraform-number-set-attribute"
export * from "./terraform-object-attribute"
export * from "./terraform-set-attribute"
export * from "./terraform-string-attribute"
export * from "./terraform-string-list-attribute"
export * from "./terraform-string-map-attribute"
export * from "./terraform-string-set-attribute"
