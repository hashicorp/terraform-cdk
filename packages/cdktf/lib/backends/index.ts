export * from "./s3-backend";
export * from "./local-backend";
export * from "./remote-backend";
export * from "./artifactory-backend";
export * from "./azurerm-backend";
export * from "./consul-backend";
export * from "./cos-backend";
export * from "./etcd-backend";
export * from "./etcdv3-backend";
export * from "./gcs-backend";
export * from "./http-backend";
export * from "./manta-backend";
export * from "./oss-backend";
export * from "./pg-backend";
export * from "./swift-backend";
export * from "./cloud-backend";
