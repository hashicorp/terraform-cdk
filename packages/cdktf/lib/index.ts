export * from './tokens'
export * from './terraform-stack';
export * from './terraform-element';
export * from './terraform-module';
export * from './terraform-provider';
export * from './terraform-resource';
export * from './terraform-output';
export * from './complex-computed-list';
export * from './resource'
export * from './testing';
export * from './app';