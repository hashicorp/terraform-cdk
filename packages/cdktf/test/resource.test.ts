
import { Testing, TerraformStack } from '../lib';
import { TestProvider, TestResource, OtherTestResource } from './helper'

test('minimal configuration', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new TestResource(stack, 'test', {
    name: 'foo'
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('with provider alias', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new TestProvider(stack, 'provider', {
    accessKey: '1'
  })

  const provider = new TestProvider(stack, 'aliased-provider', {
    alias: "foo",
    accessKey: '123'
  })

  new TestResource(stack, 'test', {
    name: 'bar',
    provider
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('provider setter', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  const provider = new TestProvider(stack, 'provider', {
    accessKey: '1'
  })

  const resource = new TestResource(stack, 'test', {
    name: 'bar'
  });

  resource.provider = provider;

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('resource fqn', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  const resource = new TestResource(stack, 'test', {
    name: 'bar'
  });

  expect(resource.fqn).toEqual('test_resource.test_DBA94737')
})

test('serialize list interpolation', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'tests');

  const resource = new TestResource(stack, 'test', {
    name: 'bar'
  });

  const otherResource = new OtherTestResource(stack, 'othertest', {});
  resource.names = otherResource.names

  expect(Testing.synth(stack)).toMatchSnapshot();
});

