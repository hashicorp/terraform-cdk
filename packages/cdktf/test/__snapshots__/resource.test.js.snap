// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal configuration 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"test_resource\\": {
      \\"test_DBA94737\\": {
        \\"name\\": \\"foo\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test\\",
            \\"uniqueId\\": \\"test_DBA94737\\"
          }
        }
      }
    }
  }
}"
`;

exports[`provider setter 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"provider\\": {
    \\"test\\": [
      {
        \\"access_key\\": \\"1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"test_resource\\": {
      \\"test_DBA94737\\": {
        \\"name\\": \\"bar\\",
        \\"provider\\": \\"test\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test\\",
            \\"uniqueId\\": \\"test_DBA94737\\"
          }
        }
      }
    }
  }
}"
`;

exports[`serialize list interpolation 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"tests\\"
    }
  },
  \\"resource\\": {
    \\"test_resource\\": {
      \\"tests_test_A231270F\\": {
        \\"name\\": \\"bar\\",
        \\"names\\": \\"\${other_test_resource.tests_othertest_E4B4389F.names}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"tests/test\\",
            \\"uniqueId\\": \\"tests_test_A231270F\\"
          }
        }
      }
    },
    \\"other_test_resource\\": {
      \\"tests_othertest_E4B4389F\\": {
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"tests/othertest\\",
            \\"uniqueId\\": \\"tests_othertest_E4B4389F\\"
          }
        }
      }
    }
  }
}"
`;

exports[`with provider alias 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"provider\\": {
    \\"test\\": [
      {
        \\"access_key\\": \\"1\\"
      },
      {
        \\"access_key\\": \\"123\\",
        \\"alias\\": \\"foo\\"
      }
    ]
  },
  \\"resource\\": {
    \\"test_resource\\": {
      \\"test_DBA94737\\": {
        \\"name\\": \\"bar\\",
        \\"provider\\": \\"test.foo\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test\\",
            \\"uniqueId\\": \\"test_DBA94737\\"
          }
        }
      }
    }
  }
}"
`;
